docs/manual-EN/1.overview/3.design-and-architecture/2.storage-design.md

```diff
@@ -1,2 +1,141 @@
+
+One Nebula kv store cluster supports multiple spaces, and each space is set with its own partition number and replica number. Different spaces are isolated physically and spaces in the same clusters supports different storage engine and sharding strategy.
+
+## Raft
+
+This part gives some details on how raft works for Nebula.
+
+### Multi Raft Group
+
+As the log IDs must be serial, almost all the implementations adapt Multi Raft Group to solve this problem. Therefore, partition number almost determines the performance of the Raft Group,but it doesn't mean the more partitions, the better the performance. Each raft group stores a set of state information and has its own WAL file, thus more partitions requires more costs. Also, when there are too many Partitions, the batch operation is meaningless if the load is not high enough. For example, a single online system with 1w tps, if its partition number is greater than ten thousand, the tps of each partition per second is likely only 1, so the batch operation loses its meaning and increases CPU cost.
+
+There are two key points to achieve the Multi Raft Group. **The first is to share the transport layer**, because each Raft Group needs to send messages to the corresponding peers. If the transport layer cannot be shared, the connection costs are huge. **Second is the threading model**, multi Raft Group must share a set of thread pools, otherwise will end with too many threads in the system, resulting in a large amount of context switch cost.
+
+### Batch
+
+For each Partition, it is necessary to do batch to improve throughput when writing WAL serially. In general, there is nothing special about batch, but Nebula designs some special types of WAL based on each part serialization, which brings some challenges.
+
+For example, Nebula uses WAL to implement lock-free CAS operations, and each CAS operation will execute after all the previous WAL are committed . So for a batch, if there are several WALs in CAS type, we need to divide this batch into several smaller serial groups.
+
+### Learner
+
+New machines need to "catch up" data for quite a long time, and there may be accidents during the process. If a new machine join the raft group as a follower, it will reduce the HA capacity of the entire cluster. Nebula uses the command WAL mentioned above to implement its learner. If the leader encounters the add learner command when writing WAL, it will add the learner to its peers and mark it as learner, so even if leaner is not included when counting majority, logs will send to them as normal. Of course, learner will not start election automatically.
```

Comment:
If the leader encounters the `add learner command` when writing WAL, it will add the learner to its peers and mark it as a learner. So logs will send to them as normal. But the learners will not evolve in the leader's election.